rfcs/SubschemaComposability.md

```diff
@@ -0,0 +1,239 @@
+Inversely, it may be impossible to generate a valid composite schema containing
+all of the GraphQL elements of a given set of subschemas; this set of subschemas
+is considered `not composable`.
+
+For a given set of subschemas and a given set of overlapping GraphQL elements
+(e.g. types, fields, arguments, etc.), it may be possible to generate a valid
+`composite` GraphQL element containing all of the overlapping GraphQL elements'
+constituent GraphQL sub-element across all of the individual subschemas. This
+GraphQL element is `composable` across the given set of overlapping elements. A
+set of subschemas is `composable` if and only if each of its overlapping GraphQL
+elements are `composable`.
+
+# Types of Composition
+
+Many types of schema composition exist, with potential distinctions including
+(but not limited to):
+
+## Timing of Composition
+
+1. "Build-time" composition in which the composite schema is created during a
+   build process prior to system start-up.
+2. "Run-time" composition in which the composite schema is created dynamically
+   at system start-up.
+3. "Dynamic" composition , in that subschemas are polled by a managing service
+   (or report changes to a managing service), such that the composition process
+   can be triggered at or after system start-up and that the composite schema is
+   dynamically refreshed as the subschemas change (or go offline!).
+
+## Method of Composite Schema Execution
+
+1. `Subschemas as remote GraphQL services`: in certain forms of composition
+   (gateways, gateway schemas, etc.), the execution of a request against a
+   composite schema will yield request(s) against spec-compliant (or even
+   non-spec compliant) GraphQL subservices that are then merged into a single
+   request.
+2. `Single-service composite schema execution`: Often, especially when utilizing
+   "build-time" composition, a single service executes the composite schema
+   without delegation to remote services tied to each subschema. Subschemas
+   exist for the purpose of code organization, but only the composite schema
+   exists at the time of execution.
+
+The below RFC is meant to be useful across all varieties of schema composition.
+Because all forms of schema composition require a composite schema, we can speak
+of "composability" without referring to the timing of composition or the type of
+execution. However, practically speaking, the method of composite schema
+execution matters significantly (much more than the timing of composition). For
+example, if a request against the composite schema must ultimately be translated
+to request(s) against the still extant source subschemas, then the composition
+must be -- in some sense -- reversed during execution, which is not always
+possible, and introduces significant complexity.
+
+The below RFC examines composability separately for the above two forms of
+composite schema execution.
+
+# Techniques for Maintaining Composability
```

Comment:
We may want to also consider developing some new techniques for maintaining composability, too. I recall we discussed the issue where schemas could be composed but the schema owners have no idea it is being composed, which creates problems when it becomes incompatible with other subschemas in the future.

One strategy might be to explicitly "opt-in" to composition. In other words, the schema contains metadata which says "This schema is allowed to be a subschema."

Going a step further, the opt-in metadata (directive?) could indicate a specific composition strategy. For example, say I wanted to provide my schema as-is. However, if someone wanted to use my schema as a subschema, then I could include a directive that says "if you compose this, then use this namespace/transform."

Another option might be simply: The composer (e.g. registry) knows which subschemas should be composed and all are guaranteed to be opt-in. This isn't always true but it may be a decent, practical solution.

This might be better suited in the discussions tab, however I am curious what our initial thoughts are. Maybe it'd be worth discussing this week in the WG, who knows 🙂

Reply:
The problem of remote subschema breaking changes bubbling up to the composed schema is pretty serious — but is it more serious than any remote service breaking change would be for a non-composed GraphQL service?

In any event, I am not clear how subschema metadata would help us navigate the challenge. If a subschema removes a field, for example, what kind of metadata could help?

If a subschema adds a field that clashes with another subschema, presumably the composed schema has no knowledge of the change, so the clash will not actually cause a runtime error.

Standardizing subschema metadata seems useful to me, although I was mostly thinking about in terms of nomenclature for how to perform cross-service resolution at runtime, I.e. key directives…

Reply:
I should also note that robustly maintaining composability across schema changes in my mind is the job of the composition service and may in large part consist of managing a schema registry for the subschemas.

Metadata that might be useful could be a schema version, or a hash of the SDL.

Reply:
All good points. After thinking it through a bit more, the metadata may not help as much. The hope there was to standardize how a subschema wanted to be composed, but that may only be useful when trying to isolate subschemas from one another.

The backward compatibility issue is especially pertinent. Maybe we can kick the can for now, and in the future some new guidelines could recommend ways to handle those issues.

>I should also note that robustly maintaining composability across schema changes in my mind is the job of the composition service and may in large part consist of managing a schema registry for the subschemas.

I agree, the composition service should be responsible (at least in part) for safe and back-compatible merges. However, it seems how this should be done is still up in the air.